File: worker/src/main/java/com/minethehung/worker/service/JobService.java
package com.minethehung.worker.service;

import com.minethehung.common.JobKafkaDTO;
import com.minethehung.worker.entity.Job;
import com.minethehung.worker.entity.JobStatus;
import com.minethehung.worker.kafka.JobProducer;
import com.minethehung.worker.repository.JobRepo;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class JobService {
    private final JobRepo jobRepo;
    private final RandomService randomService;
    private final JobProducer jobProducer;

    @Transactional
    public void saveJob(JobKafkaDTO jobKafkaDTO) {
        var job = jobRepo.findById(jobKafkaDTO.id()).orElse(null);
        if (job != null && job.getStatus() != JobStatus.SUCCESS && job.getStatus() != JobStatus.PROCESSING) {
            job.setStatus(JobStatus.PROCESSING);
            executeJob(jobRepo.saveAndFlush(job), jobKafkaDTO);
        }
    }

    private void executeJob(Job job, JobKafkaDTO jobKafkaDTO) {
        System.out.println("Executing job " + job.getId().toString() + ", retry: " + job.getRetryCount());
        Boolean status = randomService.genJobStatus();
        if (status) {
            System.out.println("Executed job: " + job.getId() + "\n");
            job.setStatus(JobStatus.SUCCESS);
        }
        else {
            System.out.println("FAIL to Executed job: " + job.getId() + "\n");
            job.setStatus(JobStatus.FAILED);
            if (job.getRetryCount() == 0) {
                job.setRetryCount(job.getRetryCount() + 1);
                jobProducer.sendJob(jobKafkaDTO, "job-topic-retry-2s");
            }
            else if (job.getRetryCount() == 1) {
                job.setRetryCount(job.getRetryCount() + 1);
                jobProducer.sendJob(jobKafkaDTO, "job-topic-retry-5s");
            }
            else if  (job.getRetryCount() == 2) {
                job.setRetryCount(job.getRetryCount() + 1);
                jobProducer.sendJob(jobKafkaDTO, "job-topic-retry-15s");
            }
            else {
                jobProducer.sendJobToDLQ(jobKafkaDTO);
            }
        }
        jobRepo.saveAndFlush(job);
    }
}

File: api-server/src/main/kotlin/com/minethehung/apiserver/service/JobService.kt
package com.minethehung.apiserver.service

import com.minethehung.apiserver.dto.JobKafkaDTO
import com.minethehung.apiserver.dto.JobRequestDTO
import com.minethehung.apiserver.entity.Job
import com.minethehung.apiserver.entity.JobStatus
import com.minethehung.apiserver.kafka.JobProducer
import com.minethehung.apiserver.repository.JobRepo
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service
import java.time.Instant
import java.util.UUID


@Service
class JobService (
    private val jobRepo: JobRepo,
    private val jobProducer: JobProducer) {
    @Transactional
    fun createJob(jobRequestDTO: JobRequestDTO): Job {
        val uuid = UUID.randomUUID()
        val payload = jobRequestDTO.payload
        val type = jobRequestDTO.type
        val create = Instant.now()
        val updated = Instant.now()
        val status = JobStatus.CREATED
        val job = Job(uuid, type, payload, 0, status, create, updated)
        return jobRepo.saveAndFlush(job)
    }

    @Transactional
    fun sendJobToKafka(job: Job): Job {
        val jobKafkaDTO = JobKafkaDTO(job.id, job.type, job.payload)
        if (jobProducer.send(jobKafkaDTO)) {
            val updatedJob = Job(job.id, job.type, job.payload, job.retryCount, JobStatus.QUEUED, job.createdAt,
                Instant.now())
            jobRepo.saveAndFlush(updatedJob)
        }
        return job
    }

    fun createAndSendJob(jobRequestDTO: JobRequestDTO): Job {
        return  sendJobToKafka(createJob(jobRequestDTO))
    }
}